compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt

```diff
@@ -0,0 +1,444 @@
+        change = event.changes[0]
+    } while (change.uptimeMillis < minUptime)
+    event
+}
+
+private suspend fun AwaitPointerEventScope.awaitReleaseOrCancelled(
+    consumeUntilRelease: Boolean = false,
+    filter: (PointerEvent) -> Boolean
+): PointerEvent? {
+    var event: PointerEvent? = null
+
+    while (event == null) {
+        event = awaitPointerEvent()
+
+        val cancelled = event.changes.fastAny {
+            it.isOutOfBounds(size, Size.Zero)
+        }
+
+        if (cancelled) return null
+
+        event = event.takeIf {
+            it.isAllPressedUp(requireUnconsumed = true) && filter(it)
+        }
+
+        if (consumeUntilRelease) {
+            currentEvent.changes.fastForEach { it.consume() }
+        }
+
+        // Check for cancel by position consumption. We can look on the Final pass of the
+        // existing pointer event because it comes after the Main pass we checked above.
+        val consumeCheck = awaitPointerEvent(PointerEventPass.Final)
+        if (consumeCheck.changes.fastAny { it.isConsumed }) {
+            return null
+        }
+    }
+
+    return event
+}
+
+private fun PointerEvent.isAllPressedDown(requireUnconsumed: Boolean = true) =
+    type == PointerEventType.Press &&
+        changes.fastAll { it.type == PointerType.Mouse && (!requireUnconsumed || !it.isConsumed) } ||
+        changes.fastAll { if (requireUnconsumed) it.changedToDown() else it.changedToDownIgnoreConsumed() }
+
+private fun PointerEvent.isAllPressedUp(requireUnconsumed: Boolean = true) =
+    type == PointerEventType.Release &&
+        changes.fastAll { it.type == PointerType.Mouse && (!requireUnconsumed || !it.isConsumed) } ||
+        changes.fastAll { if (requireUnconsumed) it.changedToUp() else it.changedToUpIgnoreConsumed() }
+
+private class PressGestureScopeImpl( // copy-pasted from TapGestureDetector. Remove when upstreaming into commonMain
+    density: Density
+) : PressGestureScope, Density by density {
+    private var isReleased = false
+    private var isCanceled = false
+    private val mutex = Mutex(locked = false)
+
+    /**
+     * Called when a gesture has been canceled.
+     */
+    fun cancel() {
+        isCanceled = true
+        mutex.unlock()
+    }
+
+    /**
+     * Called when all pointers are up.
+     */
+    fun release() {
+        isReleased = true
+        mutex.unlock()
+    }
+
+    /**
+     * Called when a new gesture has started.
+     */
+    fun reset() {
+        mutex.tryLock() // If tryAwaitRelease wasn't called, this will be unlocked.
+        isReleased = false
+        isCanceled = false
+    }
+
+    override suspend fun awaitRelease() {
+        if (!tryAwaitRelease()) {
+            throw GestureCancellationException("The press gesture was canceled.")
+        }
+    }
+
+    override suspend fun tryAwaitRelease(): Boolean {
+        if (!isReleased && !isCanceled) {
+            mutex.lock()
+        }
+        return isReleased
+    }
+}
+
+
+/*
```

Comment:
I wanted to keep them somewhere.  I remove them from the code and post it here just in case:

```
/*
 * Other notes:
 * 1) onPress? - might be useful to select a component, but usually onClick and onPress show no noticeable difference.
 * 2) onDoubleClick. It can be onDoublePress on some OS (e.g. windows). But usually the difference is not noticeable.
 * 3) no indication. users can add it along with interactionSource
 * 4) name = onClick
 *
 * clickable can use "onClick" + add indnication, + hoverable + focusable + onKey(enter) == Click
 * onContextMenuClick can use "onClick" too
 */
```

Reply:
Some of the points isn't understandable without the context. Maybe we can remove p.4 and clarify what onContextMenuClick means (or just remove it), because it isn't added yet. (and fix typo in indnication :) )

Reply:
I removed these notes from the code. Let's keep them in PR description.